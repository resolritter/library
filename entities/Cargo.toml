[package]
name = "entities"
version = "0.1.0"
authors = ["resolritter <resolritter@users.noreply.github.com>"]
edition = "2018"

[dependencies]
surf = "2.0"
serde = { version = "1", features = ["derive"] }
serde_json = "1"
