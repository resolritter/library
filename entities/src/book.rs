use crate::data::{BookEndLoanByTitlePayload, BookLeaseByTitlePayload, UserPublic};
use crate::{book_route, end_loan_route, lease_route, server_root};
use surf::{self, http::mime::JSON, RequestBuilder, Response, StatusCode};

pub fn do_borrow(
    server_addr: &str,
    user: Option<&UserPublic>,
    access_token: &str,
    payload: &BookLeaseByTitlePayload,
) -> RequestBuilder {
    let mut req = surf::post(format!(
        concat!(server_root!(), book_route!(), lease_route!()),
        server_addr, payload.title, payload.lease_id_req
    ))
    .body(serde_json::json!(payload).to_string())
    .header("X-Auth", access_token)
    .content_type(JSON);
    if let Some(user) = user {
        req = req.header("From", &*user.email)
    }
    req
}
pub async fn borrow(
    server_addr: &str,
    user: Option<&UserPublic>,
    access_token: &str,
    payload: &BookLeaseByTitlePayload,
) -> Response {
    let r = do_borrow(server_addr, user, access_token, payload)
        .await
        .unwrap();
    assert!(r.status() == StatusCode::Ok);
    r
}

pub fn do_end_loan(
    server_addr: &str,
    email: &str,
    payload: &BookEndLoanByTitlePayload,
) -> RequestBuilder {
    surf::patch(format!(
        concat!(server_root!(), book_route!(), end_loan_route!()),
        server_addr, payload.title, payload.lease_id_req
    ))
    .body(serde_json::json!(payload).to_string())
    .header("X-Auth", payload.access_token_req.clone())
    .header("From", email)
    .content_type(JSON)
}

pub async fn end_loan(
    server_addr: &str,
    email: &str,
    payload: &BookEndLoanByTitlePayload,
) -> Response {
    let r = do_end_loan(server_addr, email, payload).await.unwrap();
    assert!(r.status() == StatusCode::Ok);
    r
}

pub async fn get(server_addr: &str, id: &str) -> Response {
    let r = surf::get(format!(
        concat!(server_root!(), book_route!()),
        server_addr, id
    ))
    .await
    .unwrap();
    assert!(r.status() == StatusCode::Ok);
    r
}
