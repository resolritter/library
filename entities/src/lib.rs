mod macros;
#[macro_use]
pub mod book;
mod data;
pub mod user;

pub use crate::data::*;
