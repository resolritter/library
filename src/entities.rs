use bastion::prelude::BastionContext;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use std::sync::Arc;

pub struct App {
    pub db_pool: &'static Arc<&'static PgPool>,
    pub bastion: &'static Arc<&'static BastionContext>,
}
impl std::fmt::Debug for App {
    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Ok(())
    }
}

#[derive(Debug)]
pub struct Config {
    pub db_url: String,
}

#[derive(Clone)]
pub struct ServerState {
    pub app: &'static Arc<&'static App>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Book {
    pub id: i32,
    pub title: String,
}

#[derive(Deserialize, Debug)]
pub struct BookGet {
    pub title: String,
}

#[derive(Debug)]
pub struct BookGetMessage {
    pub channel: crossbeam_channel::Sender<Option<Book>>,
    pub payload: BookGet,
    pub app: &'static App,
}
