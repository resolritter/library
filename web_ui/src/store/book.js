import { createSlice } from "@reduxjs/toolkit"

export const initialState = { items: [] }

export default createSlice({
  name: "book",
  initialState,
  reducers: {
    addBooks: function (state, { payload: books }) {
      state.items = state.items.concat(books)
    },
  },
})
