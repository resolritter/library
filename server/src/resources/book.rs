use crate::messages::{BookGetByTitleMsg, BookLeaseByTitleMsg};
use crate::resources::ResponseData;
use crate::state::ServerState;
use entities::{
    Book, BookGetByTitlePayload, BookLeaseByTitlePayload, BookLeaseByTitleRequestBody, BookSeed,
};
use sqlx::{postgres::PgRow, Done, PgPool, Row};
use std::time::SystemTime;
use tide::{Request, StatusCode};

pub fn from_row(row: &PgRow) -> Result<Book, sqlx::Error> {
    Ok(Book {
        id: row.try_get("id")?,
        title: row.try_get("title")?,
        lease_id: row.try_get("lease_id")?,
        lease_until: row.try_get("lease_until")?,
    })
}

#[inline(always)]
async fn extract_get(req: &Request<ServerState>) -> tide::Result<BookGetByTitlePayload> {
    Ok(BookGetByTitlePayload {
        title: req.param("title")?,
    })
}
actor_response_handler::generate!(Config {
    name: get,
    actor: Book,
    response_type: Book,
    tag: GetByTitle
});

#[inline(always)]
async fn extract_lease_book(
    req: &mut Request<ServerState>,
) -> tide::Result<BookLeaseByTitlePayload> {
    let params = BookGetByTitlePayload {
        title: req.param("title")?,
    };
    let body = req.body_json::<BookLeaseByTitleRequestBody>().await?;

    Ok(BookLeaseByTitlePayload {
        title: params.title,
        lease_length: body.lease_length,
    })
}
actor_response_handler::generate!(Config {
    name: lease_book,
    actor: Book,
    response_type: String,
    tag: LeaseByTitle
});

#[inline(always)]
pub async fn get_by_title(msg: &BookGetByTitleMsg) -> Result<ResponseData<Book>, sqlx::Error> {
    let raw = sqlx::query("SELECT * FROM book WHERE title=$1")
        .bind(&msg.payload.title)
        .fetch_optional(msg.db_pool)
        .await?;

    if let Some(row) = raw {
        Ok(ResponseData(StatusCode::Ok, Some(from_row(&row)?)))
    } else {
        Ok(ResponseData(StatusCode::NotFound, None))
    }
}

#[inline(always)]
pub async fn lease_by_id(msg: &BookLeaseByTitleMsg) -> Result<ResponseData<String>, sqlx::Error> {
    let now = match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
        Ok(n) => n.as_secs(),
        Err(_) => panic!("SystemTime before UNIX EPOCH!"),
    } as i64;
    let title = &msg.payload.title;

    let book_exists = sqlx::query("SELECT EXISTS(SELECT 1 FROM book WHERE title=$1)")
        .bind(title)
        .fetch_one(msg.db_pool)
        .await?
        .try_get::<bool, usize>(0)?;

    if !book_exists {
        return Ok(ResponseData(StatusCode::NotFound, None));
    }

    let lease_length = &msg.payload.lease_length;
    let try_lease = sqlx::query(
        "UPDATE book SET lease_until=$1 WHERE title=$2 AND lease_until IS NULL OR lease_until < $3",
    )
    .bind(now + lease_length)
    .bind(title)
    .bind(now)
    .execute(msg.db_pool)
    .await?;

    if try_lease.rows_affected() == 0 {
        Ok(ResponseData(
            StatusCode::Forbidden,
            Some("This book is not available for borrowing".to_string()),
        ))
    } else {
        Ok(ResponseData(StatusCode::Ok, None))
    }
}

endpoint_actor::generate!({ actor: Book }, {
    GetByTitle: get_by_title,
    LeaseByTitle: lease_by_id,
});

pub fn seed_entities() -> [BookSeed; 3] {
    [
        BookSeed {
            title: "Cinderella".to_string(),
        },
        BookSeed {
            title: "Rapunzel".to_string(),
        },
        BookSeed {
            title: "Snow White".to_string(),
        },
    ]
}

pub async fn seed(pool: &PgPool) -> Vec<Book> {
    let seeding = seed_entities();
    let titles = &seeding
        .iter()
        .map(|b| b.title.clone())
        .collect::<Vec<String>>();

    sqlx::query("INSERT INTO book (title) VALUES (UNNEST($1::TEXT[])) RETURNING *")
        .bind(titles)
        .fetch_all(pool)
        .await
        .unwrap()
        .iter()
        .map(|b| from_row(b).unwrap())
        .collect()
}
