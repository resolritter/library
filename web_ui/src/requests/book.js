import { StatusCodes } from "http-status-codes"
import { apiEndpoints } from "src/constants"
import { store } from "src/setup"
import booksStore from "src/store/book"

import { getCors, handleErrorResponse } from "."

export const loadBooks = async function () {
  const response = await fetch(apiEndpoints.books(), {
    mode: getCors(),
  })

  if (response.status === StatusCodes.OK) {
    store.dispatch(booksStore.actions.addBooks(await response.json()))
    return { status: "ok" }
  } else {
    return await handleErrorResponse(response)
  }
}
