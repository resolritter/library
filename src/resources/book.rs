use crate::entities::{Book, BookByTitlePayload, ServerState};
use crate::messages::{ActorGroups, BookByTitleMsg, BOOK};
use crate::resources::respond_with;
use bastion::prelude::*;
use sqlx::{PgPool, Row};
use tide::{Request, Response};

#[inline(always)]
pub async fn fetch_one_by_title(pool: &PgPool, title: &str) -> Result<Option<Book>, sqlx::Error> {
    let raw = sqlx::query("SELECT * FROM Book WHERE title=$1")
        .bind(title)
        .fetch_optional(pool)
        .await?;

    if let Some(row) = raw {
        Ok(Some(Book {
            id: row.try_get("id")?,
            title: row.try_get("title")?,
        }))
    } else {
        Ok(None)
    }
}

pub async fn get(req: Request<ServerState>) -> tide::Result<Response> {
    let payload = BookByTitlePayload {
        title: req.param("title").unwrap(),
    };
    let (reply, r) = crossbeam_channel::bounded::<Option<Book>>(1);
    let state = req.state();

    unsafe {
        BOOK.get()
            .unwrap()
            .read()
            .as_ref()
            .unwrap()
            .send(BookByTitleMsg {
                reply,
                payload,
                db_pool: state.global.db_pool,
            })
            .unwrap();
    }

    crossbeam_channel::select! {
      recv(r) -> msg => {
        match msg {
          Ok(item) => match item {
            Some(book) => respond_with::<Book>(Some(book), tide::StatusCode::Ok),
            None => respond_with::<Book>(None, tide::StatusCode::NotFound),
          },
          _ => respond_with::<Book>(None, tide::StatusCode::InternalServerError)
        }
      },
      default(std::time::Duration::from_secs(3)) => respond_with::<Book>(None, tide::StatusCode::RequestTimeout),
    }
}

pub fn actor(children: Children) -> Children {
    children
        .with_name(ActorGroups::Book.as_ref())
        .with_exec(move |_| async move {
            let (channel, r) = crossbeam_channel::unbounded::<BookByTitleMsg>();
            unsafe {
                let mut lock = BOOK.get().unwrap().write();
                *lock = Some(channel);
            }
            println!("Book actor is ready!");

            loop {
                let BookByTitleMsg {
                    reply,
                    payload,
                    db_pool,
                } = r.recv().unwrap();

                let _ = reply.send(match fetch_one_by_title(db_pool, &payload.title).await {
                    Ok(output) => output,
                    err => {
                        println!("{:#?}", err);
                        None
                    }
                });
            }

            #[allow(unreachable_code)]
            Ok(())
        })
}
