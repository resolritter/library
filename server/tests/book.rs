use entities::access_mask;
use insta::assert_snapshot;
use stdext::function_name;
use surf::StatusCode;
use tempdir::TempDir;
use test_utils::{format::format_test_name, read_snapshot, spawn_test_program, SpawnedTest};

#[async_std::test]
async fn test_get() {
    use entities::book;

    let test_name = format_test_name(function_name!());
    let tmp_dir = TempDir::new(&test_name).unwrap();
    let SpawnedTest {
        server_addr,
        log_dir,
        process: _,
    } = &spawn_test_program(&tmp_dir, None);

    let mut get = book::get(&server_addr, "Rapunzel").await;
    assert_snapshot!(test_name, get.body_string().await.unwrap());

    assert_snapshot!(read_snapshot(&log_dir));
}

#[async_std::test]
async fn test_lease() {
    use entities::{book, user, BookLeaseByTitlePayload, UserCreationPayload, UserPublic};

    let test_name = format_test_name(function_name!());
    let tmp_dir = TempDir::new(&test_name).unwrap();
    let SpawnedTest {
        server_addr,
        log_dir,
        process: _,
    } = &spawn_test_program(&tmp_dir, None);

    const WHOLE_DAY: i64 = 86400;
    let (_, first_user) = user::create(
        &server_addr,
        &UserCreationPayload {
            email: "simple@user.com".to_string(),
            access_mask: access_mask::USER,
            requester_access_token: None,
        },
    )
    .await;
    let payload = BookLeaseByTitlePayload {
        title: "Rapunzel".to_string(),
        lease_id_req: first_user.email.clone(),
        lease_length: WHOLE_DAY,
    };

    // The route should be protected against guest users
    let guest_user = book::do_borrow(&server_addr, None, "DOES_NOT_MATTER", &payload)
        .await
        .unwrap();
    assert!(guest_user.status() == StatusCode::Forbidden);

    // The route should be protected against ficticious users
    let bad_ficticious_user = book::do_borrow(
        &server_addr,
        Some(&UserPublic {
            email: "DOES_NOT_EXIST".to_string(),
            access_mask: access_mask::USER,
            access_token: "DOES_NOT_MATTER".to_string(),
        }),
        &first_user.access_token,
        &payload,
    )
    .await
    .unwrap();
    assert!(bad_ficticious_user.status() == StatusCode::Forbidden);

    // The route should be protected against users with invalid tokens
    let bad_invalid_token_user =
        book::do_borrow(&server_addr, Some(&first_user), "INVALID_TOKEN", &payload)
            .await
            .unwrap();
    assert!(bad_invalid_token_user.status() == StatusCode::Forbidden);

    // Lease a book for a whole day
    assert_snapshot!(
        format!("{}_good", test_name),
        book::borrow(
            &server_addr,
            Some(&first_user),
            &first_user.access_token,
            &payload
        )
        .await
        .body_string()
        .await
        .unwrap()
    );

    // The book has been leased successfully, thus the following should not work
    let bad_borrow = book::do_borrow(
        &server_addr,
        Some(&first_user),
        &first_user.access_token,
        &payload,
    )
    .await
    .unwrap();
    assert!(bad_borrow.status() == StatusCode::Forbidden);

    assert_snapshot!(read_snapshot(&log_dir));
}

#[async_std::test]
async fn test_end_loan() {
    use entities::{
        book, user, BookEndLoanByTitlePayload, BookLeaseByTitlePayload, UserCreationPayload,
    };

    let test_name = format_test_name(function_name!());
    let tmp_dir = TempDir::new(&test_name).unwrap();
    let admin_access_token = "ADMIN";
    let admin_email = "admin@admin.com";
    let SpawnedTest {
        server_addr,
        log_dir,
        process: _,
    } = &spawn_test_program(
        &tmp_dir,
        Some(format!("{}::{}", admin_email, admin_access_token)),
    );

    const WHOLE_DAY: i64 = 86400;
    let (_, first_user) = user::create(
        &server_addr,
        &UserCreationPayload {
            email: "first@user.com".to_string(),
            access_mask: access_mask::USER,
            requester_access_token: None,
        },
    )
    .await;
    let (_, second_user) = user::create(
        &server_addr,
        &UserCreationPayload {
            email: "second@user.com".to_string(),
            access_mask: access_mask::USER,
            requester_access_token: None,
        },
    )
    .await;
    let mut borrow_payload = BookLeaseByTitlePayload {
        title: "Rapunzel".to_string(),
        lease_id_req: first_user.email.clone(),
        lease_length: WHOLE_DAY,
    };

    // Borrow a book with the first user
    book::borrow(
        &server_addr,
        Some(&first_user),
        &first_user.access_token,
        &borrow_payload,
    )
    .await;

    // The second user won't be able to end the loan on behalf of the first
    let mut end_loan_payload = BookEndLoanByTitlePayload {
        title: borrow_payload.title.clone(),
        lease_id_req: borrow_payload.lease_id_req.clone(),
        access_token_req: second_user.access_token.to_string(),
    };
    let bad_forbidden_end_loan =
        book::do_end_loan(&server_addr, &second_user.email, &end_loan_payload)
            .await
            .unwrap();
    assert!(bad_forbidden_end_loan.status() == StatusCode::Forbidden);

    // The first user can end the loan on its own
    end_loan_payload = BookEndLoanByTitlePayload {
        access_token_req: first_user.access_token.to_string(),
        ..end_loan_payload
    };
    book::end_loan(&server_addr, &first_user.email, &end_loan_payload).await;

    // Now the book is free and the second user will be able to borrow it
    borrow_payload = BookLeaseByTitlePayload {
        lease_id_req: second_user.email.clone(),
        ..borrow_payload
    };
    book::borrow(
        &server_addr,
        Some(&second_user),
        &second_user.access_token,
        &borrow_payload,
    )
    .await;

    // Librarians will be able to end the loan on the behalf of normal users
    let (_, librarian_user) = user::create(
        &server_addr,
        &UserCreationPayload {
            email: "librarian@user.com".to_string(),
            access_mask: access_mask::LIBRARIAN,
            requester_access_token: Some(admin_access_token.to_string()),
        },
    )
    .await;
    book::end_loan(
        &server_addr,
        &librarian_user.email,
        &BookEndLoanByTitlePayload {
            title: borrow_payload.title.clone(),
            lease_id_req: borrow_payload.lease_id_req.clone(),
            access_token_req: librarian_user.access_token.to_string(),
        },
    )
    .await;

    assert_snapshot!(read_snapshot(&log_dir));
}
