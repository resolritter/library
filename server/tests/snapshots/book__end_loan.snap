---
source: server/tests/book.rs
expression: read_snapshot(&log_dir)
---
UserCreationPayload {
    email: "first@user.com",
    access_mask: 1,
    requester_access_token: None,
}
UserCreationPayload {
    email: "second@user.com",
    access_mask: 1,
    requester_access_token: None,
}
BookLeaseByTitlePayload {
    title: "Rapunzel",
    lease_id_req: "first@user.com",
    lease_length: 86400,
}
BookEndLoanByTitlePayload {
    title: "Rapunzel",
    lease_id_req: "first@user.com",
    access_token_req: "second@user.com",
}
BookEndLoanByTitlePayload {
    title: "Rapunzel",
    lease_id_req: "first@user.com",
    access_token_req: "first@user.com",
}
BookLeaseByTitlePayload {
    title: "Rapunzel",
    lease_id_req: "second@user.com",
    lease_length: 86400,
}
UserCreationPayload {
    email: "librarian@user.com",
    access_mask: 17,
    requester_access_token: Some(
        "ADMIN",
    ),
}
BookEndLoanByTitlePayload {
    title: "Rapunzel",
    lease_id_req: "second@user.com",
    access_token_req: "librarian@user.com",
}

